### DISCUSSION TITLE
将便携版vapoursynth（python）与mpv目录分开的新方法

### DISCUSSION THREAD

**Discussion by zhongfly** [hooke007/mpv_PlayKit#484] [mpv 通用]
一般来说mpv使用vapoursynth都需要把VapourSynth Portable版解压到mpv.exe的同一目录里，并且还常搭配便携版python一起使用。这么一来mpv文件夹中就变得相当混乱了，对于有洁癖的用户可能难以忍受。

我的方法：将下面压缩包里的VSScript.dll放在mpv.exe旁边，新建一个`VapourSynth`文件夹，把VapourSynth还有便携版Python都放在这个文件夹里（vs插件的dll也要放到`VapourSynth`文件夹里面的`vs-plugins`中），按照VapourSynth文档的说明安装。已在windows 10/11 64位系统 + mpv 64位 + VapourSynth R69 64位 + 便携版python 3.12 上测试使用svp 4，至少stats上显示正常，mpv没有报错退出
[VSScript.zip](https://github.com/user-attachments/files/16237156/VSScript.zip)
ps：我此处提供的VSScript.dll是64位的，32位系统的就不要试了。

效果如图：
![image](https://github.com/user-attachments/assets/eb8f296b-7cec-424b-addd-898481e5a24f)
![image](https://github.com/user-attachments/assets/58630493-9eca-4e8b-9c4f-a721f3cf7e65)

在Windows目前版本的mpv中（基于shinchiro/mpv-winbuild-cmake的编译工具链），当mpv需要使用VapourSynth功能时，会先在同一目录下寻找VSScript.dll，通过这个dll来调用VapourSynth，而VapourSynth再去寻找python运行环境。
我从dll劫持的相关文章中发现可以使用转发的方法来实现：让mpv调用一个定制的VSScript.dll，这个定制的VSScript.dll会再去我们预先设定的位置调用真正的VSScript.dll。
简单理解就是：用一个同名的dll文件中转调用

由于我不会这些，所以借助了现成的项目 https://github.com/itisluiz/DLLForward ， 略作修改，使生成的VSScript.dll会转发到相对路径`./VapourSynth/`里的VSScript.dll
源码： [VSScript-source.zip](https://github.com/user-attachments/files/16237137/VSScript-source.zip)
使用visual studio 2022编译。如果想要修改成其他路径的，自行修改VSScript.h的188行

**Comment by verygoodlee** [comment]
这样弄在单独使用python时是不是有问题，便携版python不止vapoursynth需要用到，
lua脚本里也可能使用subprocess调用python，应该也是优先在mpv根目录寻找，然后去环境变量找

**Reply by zhongfly** [reply]
还有lua调用python脚本的吗？这也太罕见了…我没见过这样的

如果真有需要的，也许有方法做一个python.exe的转发？

**Reply by verygoodlee** [reply]
比如这个加载B站弹幕的脚本 [MPV-Play-BiliBili-Comments](https://github.com/itKelis/MPV-Play-BiliBili-Comments/blob/b71fa74532c9daddc1606b9b1b86c6e9c519609f/scripts/bilibiliAssert/main.lua#L110-L142)

**Reply by zhongfly** [reply]
最简单的办法，就是稍微改一下lua脚本，让它多搜索一个路径有没有python.exe，或者增加一个选项可以指定python路径…

**Reply by zhongfly** [reply]
看了下，exe这类可执行文件的检索逻辑似乎更容易利用，不需要修改脚本。运行命令python时，会先在当前目录搜索python.com，再搜python.exe，再搜python.bat，等等，只有当前目录里找不到，才会去检索环境变量path。所以做一个名字python.bat的脚本转发就行

**Reply by verygoodlee** [reply]
顺序其实是环境变量PATHEXT定义的，用bat脚本转发应该可行
![image](https://github.com/user-attachments/assets/7c7faf09-c29b-4b41-8efe-4e58f360e6ae)

**Reply by zhongfly** [reply]
python.bat脚本内容大致如下：
```bat
.\VapourSynth\python.exe %*
```
应该可以应付大部分情况

**Comment by nyfair** [comment]
我用的安装版python，只要确保python和VSScript所在路径在PATH环境变量里，vapoursynth.cp312-win_amd64.pyd所在路径在PYTHONPATH环境变量里就足够了吧
同理，都有了python环境了，yt-dlp之类的工具用脚本版足够了，没必要用自带python环境的打包版

**Reply by zhongfly** [reply]
> yt-dlp之类的工具用脚本版

可能不行吧，ytdl_hook.lua在Windows上都会搜索.exe后续的yt-dlp

https://github.com/mpv-player/mpv/blob/e509ec0aaffce74e520702e16e3e21ea0f168940/player/lua/ytdl_hook.lua#L951-L953

update: 如果找不到的话会直接用不带后缀yt-dlp，这么看来只要把yt-dlp安装为全局工具或其他类似的话就可以

**Reply by zhongfly** [reply]
另外我不用安装版python，用便携版python搭配vs的原因是避免python的版本混乱，环境变量里的python只用于自己写py用(可以用pyenv-win切换版本)

**Comment by zaeboba** [comment]
can make this file to be in portable_config? and the VapourSynth folder there as well?

**Reply by zhongfly** [reply]
If VSScript.dll is not in the same directory as mpv.exe, you need to add the folder where VSScript.dll is located to the system variable `PATH`. In this case, you don't need the modified dll file, just use the original one!

**Reply by zaeboba** [reply]
Sorry, I'm not a programmer, I just wanted to go with my portable_config folder on different computers and not to configure anything unnecessary ((

**Reply by zhongfly** [reply]
As I know, the order of searching for `VSScript.dll` is to search for its existence in the same directory first, and then search for its existence in the path in the system variable `PATH`. So without configuring any parameters, you can only put `VSScript.dll` in the same directory as `mpv.exe`. 
If you just want to put the VapourSynth folder into portable_config, just compile a `VSScript.dll` yourself using the method I described.

**Reply by zaeboba** [reply]
I must have translated it wrong, I meant it like this.
/
├── portable_config
│   ├── cache
│   │   └── other
│   ├── fonts
│   │   └── orthe
│   ├── script-modules
│   │   └── other
│   ├── script-opts
│   │   └── other
│   ├── scripts
│   │   └── other
│   ├── shaders
│   │   └── other
│   ├── vs
│   │   └── other
│   ├── VapourSynth
│   │   └── other
│   ├── yt-dlp.exe
│   ├── mpv.conf
│   ├── input.conf
│   └── VSScript.dll
└── mpv.exe

**Reply by zhongfly** [reply]
When not modifying the mpv code and not configuring any environment variables, VSScript.dll must be in the same level folder as mpv.exe.

**Reply by zaeboba** [reply]
> If you just want to put the VapourSynth folder into portable_config, just compile a `VSScript.dll` yourself using the method I described.

I got it, buddy. But I'm not a programmer and I don't know how to compile.

**Reply by zhongfly** [reply]
In short:
1. VSScript.dll in portable_config without configuring environment variables: can't
2. VapourSynth folder in portable_config: compile a VSScript.dll like this post

**Reply by zhongfly** [reply]
I can compile one for you if I have time.

**Reply by zhongfly** [reply]
@zaeboba try this for `./portable_config/VapourSynth/`
[VSScript(portable_config).zip](https://github.com/user-attachments/files/17606710/VSScript.portable_config.zip)

**Reply by zaeboba** [reply]
Thank you very much! 
I spent almost 2 hours trying to do it myself, but I never figured out how to make a .h .dll 😂
I asked chatgpt for help, but all “tips” only gave me errors. 🤣

**Reply by zhongfly** [reply]
![image](https://github.com/user-attachments/assets/5e915722-d49c-46af-be0a-6a69a5e0df24)
I have said in the end: modify VSScript.h line 188 and complie it with visual studio(open source code in vs,and `build` -> `build vsscript`).
read carefully

**Reply by zaeboba** [reply]
Yeah, I know I was editing.
but I couldn't pull myself together.
![изображение](https://github.com/user-attachments/assets/95b4aa09-0c60-4ee6-a471-70c990db7b84)

**Reply by zhongfly** [reply]
oh,You want to use g++ to compile it, and you didn't install g++.Visual studio(not visual studio code) is recommended.

**Reply by zaeboba** [reply]
Like I said before, I'm not a programmer, it's really hard for me to figure this out.

thanks again

**Comment by MiKoto-Railgun** [comment]
按上面操作后，显示依赖错误

![image](https://github.com/user-attachments/assets/d61986b1-2a6c-4186-bfa0-2b5ab85f9ff1)
![image](https://github.com/user-attachments/assets/7c0ec4e9-20a6-4cde-bb85-e3f0b3438866)

请问这个该怎么办？

**Reply by zhongfly** [reply]
和这个话题无关，python/vapoursynth能正常运行就说明不是这里的问题。
你自己按照K7sfunc和mvt的要求去检查依赖

**Reply by MiKoto-Railgun** [reply]
好的

**Reply by zhongfly** [reply]
https://hooke007.github.io/unofficial/mpv_start.html#mvtools
mvt依赖的安装，不过文档没更新新版vs的插件位置。现在应该要把libmvtools.dll放到`VapourSynth\vs-plugins`文件夹里

参考文档，新版vs插件应放在路径`<base path (portable.vs location)>\vs-plugins`：https://www.vapoursynth.com/doc/installation.html#windows-portable
对于本话题的情况，应该要把libmvtools.dll放到`\VapourSynth\vs-plugins`文件夹里
